### PR TITLE
feat(posfilter): Add Kalman Filtering functionality

### LINKED ISSUE #235: feat(posfilter): Add Kalman Filtering functionality
Integrate Kalman filtering into GNATSS. 
Rewrite algorithm to be more efficient than the implementation in PR #232

### PR DESCRIPTION
**Pull request by madhavmk**
## Overview

Added `kalman_filtering` operation as proposed by @SquirrelKnight on https://github.com/seafloor-geodesy/gnatss/pull/167. This integration includes re-writing the proposed functionality to a more functional programming with numba to really speed up the performance. A few tweaks has also been made to match the steps of the kalman filtering to the equivalent `kfwgpos.F` Fortran implementation.

Closes #235

### CODE REVIEW

File: src/gnatss/loaders.py [hunk 1]
```diff
@@ -258,42 +260,60 @@ def load_gps_solutions(
         The list of path string to files to load
     time_round : int
         The precision value to round the time values
+    is_novatel_l1_data : bool
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:263]: change to `is_l1_data`

File: src/gnatss/loaders.py [hunk 2]
```diff
@@ -258,42 +260,60 @@ def load_gps_solutions(
...
+        If True, read according to Novatel L1 data format (Usually named `GPS_POS_FREED`).
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:264]:
```suggestion
        If True, this is usually named `GPS_POS_FREED`.
```

File: src/gnatss/loaders.py [hunk 3]
```diff
@@ -258,42 +260,60 @@ def load_gps_solutions(
...
+        Else read from already Kalman Filtered data format. (Usually named `POS_FREED_TRANS_TWTT`) # TODO What's the proper namef or already filtered Kalman format??
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:265]:
```suggestion
```

File: src/gnatss/loaders.py [hunk 4]
```diff
@@ -258,42 +260,60 @@ def load_gps_solutions(
...
-    if isinstance(time_round, int) and time_round > 0:
-        all_gps_solutions.loc[:, constants.GPS_TIME] = all_gps_solutions[
-            constants.GPS_TIME
-        ].round(time_round)
+    if is_novatel_l1_data:
+        columns = [
+            constants.GPS_TIME,
+            "dtype",
+            *constants.GPS_GEOCENTRIC,
+            *constants.GPS_COV_DIAG,  # TODO Are sdx same as GPS_COV_DIAG?
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:292]:
```suggestion
            'sdx', 'sdy', 'sdz'
```

File: src/gnatss/loaders.py [hunk 5]
```diff
@@ -258,42 +260,60 @@ def load_gps_solutions(
...
+        gps_solutions = [
+            pd.read_csv(i, delim_whitespace=True, header=None, names=columns)
+            for i in files
+        ]
+        all_gps_solutions = pd.concat(gps_solutions).reset_index(drop=True)
+        # Round to match the delays precision
+        # TODO: Find a way to determine this precision dynamically?
+        if (
+            isinstance(time_round, int) and time_round > 0
+        ):  # TODO Should skip this step for novatel l1 data?
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:313]: don't skip, just leave it :)

File: src/gnatss/ops/posfilter.py [hunk 6]
```diff
@@ -75,3 +79,86 @@ def rotation(
...
+    x, P, K, Pp = run_filter_simulation(merged_np_array)
+
+    smoothed_results = DataFrame(
+        x.reshape(x.shape[0], -1),
+        columns=[*constants.GPS_GEOCENTRIC, *constants.GPS_COV_DIAG],
+    )
+    smoothed_results[constants.GPS_TIME] = merged_df[constants.GPS_TIME]
+    # TODO How are the GPS_COV values calculated???
+
+    return smoothed_results
```
Comment (lsetiawan) [inline src/gnatss/ops/posfilter.py:164]: Just leave to return the 6 cols, but check out the `x`,`y`,`z` against input values

File: src/gnatss/ops/posfilter.py [hunk 7]
```diff
@@ -1,11 +1,15 @@
...
+import numpy as np
 from nptyping import Float, NDArray, Shape
 from pandas import DataFrame
 from pymap3d import ecef2enu
 from scipy.spatial.transform import Rotation
 
+from gnatss import constants
 from gnatss.configs.solver import ArrayCenter
+from gnatss.loaders import load_gps_solutions
+from gnatss.ops.kalman import run_filter_simulation
```
Comment (lsetiawan) [inline src/gnatss/ops/posfilter.py:12]: Imports should be relative within a package in case package name changes. Please modify to be relative.

**Comment by madhavmk** [comment]
Hello @lsetiawan,
I made the changes suggested, and incorporated the diagonal values to retrieve the covarianec values. However the cov_xx, cov_yy, and the cov_zz are not close to the gold standard values in the POS_FREED_TRANS_TWTT files.
Could you take a look? (at the `Optimized_Kalman.ipynb` in particular)

**Pull request merged**